logprep/framework/rule_tree/rule_tree.py

```diff
@@ -212,8 +213,9 @@ def get_matching_rules(
             if child.does_match(event):
                 current_node = child
 
-                if current_node.matching_rules:
-                    matches += child.matching_rules
+                for matching_rule in current_node.matching_rules:
+                    if matching_rule not in matches:
+                        matches.append(matching_rule)
```

Comment:
not needed if we declare `matches` as `set`  in line 210

Reply:
This makes the rule application order random (since matches are being processed in a random order), which is no big deal in itself, but it can make problems with some tests, since they do sometimes succeed and sometimes they don't.
An example for this would be the selective extractor, since it adds labels into a list, which might then be inserted in a different order.
I imagine similar problems could occur with auto-tests. Fixing it in a unit test is relatively easy, but I'm not sure about auto-tests.
Sets might be slightly faster in this situation however.
Not sure what is more important though and if it's worth the hassle.

Reply:
then the tests have be fixed. the implementation should not rely on the order in a list. I think we should do it.
In tests we have this nice `load_specific_rule` method, which ensures that only this rule is loaded. this should help.

Reply:
I've changed it to use set again. We have to be more careful when writing tests now. Some tests don't load rules directly, but load them when creating a processor. Maybe they can be rewritten at some point to all use `load_specific_rule` instead. Auto-tests are unaffected, since they always apply one rule at a time. I will need to test this more though before I open the pull request.

Reply:
thanks.... and I agree, that we should rewrite the tests to use the `load_specific_rule` helper method, because it increases readability of the tests a lot. but we do it, when we see, that something does not work reliable.